### DISCUSSION TITLE
Is there a way to change a type definition using API?

### DISCUSSION THREAD

**Discussion by saruman9** [Vector35/binaryninja-api#3006] [API] [answered]
I have a type definition for example:

```python
>>> old_t = [t for t in bv.types if "std::__cxx11::string" == '::'.join(t[0].name)][0]
>>> old_t
('std::__cxx11::string', <type: immutable:NamedTypeReferenceClass 'typedef std::__cxx11::string'>)
```

Yes, it's a cycle type (`typedef std::__cxx11::string std::__cxx11::string;`), but it's a example. I don't use `get_type_by_name` method, because it's don't work for C++ types. 

I tried to change the type definition:

```python
>>> t = [t for t in bv.types if "std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >" == '::'.join(t[0].name)][0]
>>> t
('std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >', <type: immutable:StructureTypeClass 'class'>)
>>> new_t = NamedTypeReferenceType.create_from_type("std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >", type = t[1])
>>> new_t
<type: immutable:NamedTypeReferenceClass 'class std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >'>
>>> bv.define_type(old_t[1].type_id, "std::__cxx11::string", new_t)
'std::__cxx11::string'
```

But nothing happens. If I will undefine `std::__cxx11::string` using UI, than Binary Ninja will crash (segmentation fault).

How I can change a type definition using API?

**Comment by fuzyll** [comment, marked as answer]
Here's the documentation for that: https://docs.binary.ninja/guide/type.html#type-modification

But, you should *never* get a crash from undefining a type. I'm going to turn this into an issue to track that crash.

**Reply by saruman9** [reply]
I have already tried to modify the type definition using `mutable_copy`, but I don't know how I can modify `typedef`, because neither [NamedTypeReferenceBuilder](https://api.binary.ninja/binaryninja.types-module.html#binaryninja.types.NamedTypeReferenceBuilder) nor [TypeBuilder](https://api.binary.ninja/binaryninja.types-module.html#binaryninja.types.TypeBuilder) has the needed methods.

**Reply by fuzyll** [reply]
Yep, you're right - we're missing those in the API and will need to expose them. Thanks for pointing it out! The issue is #3014 and we'll track progress in there. Sorry for the issue.

EDIT: Re-marking this as answered because it _will_ be the right answer once we fix the issue.
